diff --git a/src/MempoolStatus.cpp b/src/MempoolStatus.cpp
index b55b312..e5ad3d1 100644
--- a/src/MempoolStatus.cpp
+++ b/src/MempoolStatus.cpp
@@ -96,7 +96,7 @@ MempoolStatus::start_mempool_status_thread()
 bool
 MempoolStatus::read_mempool()
 {
-    rpccalls rpc {deamon_url, login};
+    rpccalls rpc {daemon_url, login};
 
     string error_msg;
 
@@ -167,10 +167,10 @@ MempoolStatus::read_mempool()
         last_tx.tx = tx;
 
         // key images of inputs
-        vector<txin_to_key> input_key_imgs;
+        vector<txin_v> input_key_imgs;
 
         // public keys and xmr amount of outputs
-        vector<pair<txout_to_key, uint64_t>> output_pub_keys;
+        vector<pair<txout_target_v, uint64_t>> output_pub_keys;
 
         // sum xmr in inputs and ouputs in the given tx
         const array<uint64_t, 4>& sum_data = summary_of_in_out_rct(
@@ -222,7 +222,7 @@ MempoolStatus::read_mempool()
 bool
 MempoolStatus::read_network_info()
 {
-    rpccalls rpc {deamon_url, login};
+    rpccalls rpc {daemon_url, login};
 
     COMMAND_RPC_GET_INFO::response rpc_network_info;
 
@@ -325,8 +325,8 @@ MempoolStatus::is_thread_running()
     return is_running;
 }
 
-bf::path MempoolStatus::blockchain_path {"/home/mwo/.bitmonero/lmdb"};
-string MempoolStatus::deamon_url {"http:://127.0.0.1:18081"};
+bf::path MempoolStatus::blockchain_path {"~/.haven/lmdb"};
+string MempoolStatus::daemon_url {"http:://127.0.0.1:17750"};
 cryptonote::network_type MempoolStatus::nettype {cryptonote::network_type::MAINNET};
 atomic<bool>       MempoolStatus::is_running {false};
 boost::thread      MempoolStatus::m_thread;
